### PR TITLE
Master/Minion auto-discovery

### PR DESCRIPTION
**Pull request by isbm**
### What does this PR do?

Allows your Minion find out Master(s) without configuration, automatically finding them out on the network.

### Previous Behavior

Typical Salt Minion configuration required:
```
master: your.master.somewhere
```

### New Behavior
Possible configuration on Salt Minion:
```
```

:wink: 

### Additional Features

Suppose you have three Masters, but all of them are serving for different isolated segments in parallel. They are dockers or temporary VMs. And your network is entirely dynamic, so you don't know what IP will appear on what docker or VM for your masters after their restart. But you still want to _segregate_ these Masters the way Minions find their "home" again, after restart all that. So your Minion is connecting not to a specific IP or DNS, but to a _description_. An example of such configuration as follows:

Configuration on Master: `/etc/salt/master`:
```yaml
discovery:
  mapping:
    product: SUSE Manager
```

Configuration on Minion: `/etc/salt/minion`:
```yaml
discovery:
  match: all
  mapping:
    product: SUSE Manager
```

In this case a key `product` matches string `SUSE Manager`. The `mapping` is just merely a key/value pair of _absolutely anything_ you want. It can be `name: Steven` or `fruit: grapes` or whatever you want.

### ToDO (right next)

Right now Minion connects to a several same Masters (multi-master) by random factor. How about found list of valid masters would report their load/numbers of minions, so minion choses the least loaded one?

### Tests written?

Yes

@thatch45 @cachedout @gtmanfred Merry Christmas! :wink:

### CODE REVIEW

**Comment by isbm** [comment]
@cachedout @Ch3LL  any chance looking at this before Dec 1 freeze-up? Tests are failing for kubernetes, i.e unrelated. Code climate is complaining about overridden standard Python library method. Lint is fixed. Unit tests are there and passing.

Anything else needs to be done? Thanks!

**Comment by thatch45** [comment]
re-run ubuntu

**Comment by isbm** [comment]
@rallytime I was instructed to ping you after tests passes (done, seems like) and right after the branch is up to date sync'ed, so this can go in. :wink:

**Comment by rallytime** [comment]
re-run lint

**Comment by isbm** [comment]
@rallytime re-based again.

**Comment by isbm** [comment]
@rallytime @cachedout I forgot how many times I rebased this already. Tests passes. What else needs to be done here?

File: doc/topics/releases/oxygen.rst [hunk 1]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
 * ``iscsi_iqn``: Show the iSCSI IQN name for a host
 * ``swap_total``: Show the configured swap_total for Linux, *BSD, OS X and Solaris/SunOS
 
+Salt Minion Autodiscovery
+------------------------
+
+Salt Minion now no longer need to be configured against a specifig DNS name or IP address of a Master.
+
+For this feature Salt Master is now require port 4520 for UDP broadcast packet to be opened
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: s/is now require/now requires/

Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `broadcast packets`

File: doc/topics/releases/oxygen.rst [hunk 2]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
...
+and Salt Minion is expected to be able to send to the same port UDP packets.
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `and the Salt Minion`
`be able to send UDP packets to the same port.`

File: doc/topics/releases/oxygen.rst [hunk 3]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
...
+
+Connection to a type instead of DNS
+===================================
+
+By now each Minion was connecting to a Master by DNS or IP address. From now on it is possible
+also to connect to a _type_ of a Master. For example, in the network there are three different
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `in a network`

File: doc/topics/releases/oxygen.rst [hunk 4]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
...
+Masters, each corresponds for a particular niche or environment or specific role etc. Minion
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `role, etc.`
`The minion is supposed to connect`

File: doc/topics/releases/oxygen.rst [hunk 5]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
...
+supposed to connect only to one of those Masters that is described approriately.
+
+To achieve such effect, each `/etc/salt/master` configuration should have a `discovery` option,
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `such an effect`

File: doc/topics/releases/oxygen.rst [hunk 6]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
...
+which should have `mapping` element with arbitrary key/value pairs. The same configuration shoul
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `should have a`

File: doc/topics/releases/oxygen.rst [hunk 7]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
...
+
+Notice `match` criteria is set to `all`. This would mean that from all found Masters select only
+that, which `description` is set to `SES 5.0` _and_ `node` is set to `1`. All other Masters will
+be ignored.
+
+
+Limitations
+===========
+
+This feature has a couple of _temporary_ limitations that are subject to change in nearest future:
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `change in the future`.

File: doc/topics/releases/oxygen.rst [hunk 8]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
...
+
+- Only one Master on the network is supported. Currently Minion cannot select which Master
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `Currently the Minion`

File: doc/topics/releases/oxygen.rst [hunk 9]
```diff
@@ -95,6 +95,63 @@ The new grains added are:
...
+  out of few the same to choose. This will change to choosing Master that is least loaded.
```
Comment (cachedout) [inline doc/topics/releases/oxygen.rst]: `choosing the Master`

File: salt/config/__init__.py [hunk 10]
```diff
@@ -1144,6 +1144,30 @@ def _gather_buffer_space():
...
+    # especially when masters are not related to each other.
+    'mapping': dict,
+
+    # SSDP discovery mapping matcher policy
+    # Values: "any" where at least one key/value pair should be found or
+    # "all", where every key/value should be identical
+    'match': str,
+
+    # Port definition.
+    'port': int,
```
Comment (cachedout) [inline salt/config/__init__.py]: I am concerned that we are using the key of `port` here as a top-level key in this dictionary where it really only applies to the auto-discovery feature. I think we should prefix all of these new keys with `discovery_`

**Review by cachedout — CHANGES_REQUESTED** [review]

File: salt/config/__init__.py (hunk 10 above)
Comment (isbm) [reply salt/config/__init__.py]: Uhm... As I understand, this is just a key for whatever. So it is always like:
```yaml
discovery:
    port: 1234
```
What you suggesting is to have:
```yaml
discovery:
    discovery_port: 1234
```
I don't think this is any nice.

Comment (isbm) [reply salt/config/__init__.py]: @cachedout but if you thing that discovery is using `port` at the top level, then no way. I would never design something like that. :laughing: However, I thought that the `port` as a definition on its own can be happily used anywhere else by anything else in the future:
```yaml
some_new_stuff:
    port: 12345

pancake_burner:
    port: 12346

discover:  # as it is now
    port: 12347
```
...etc.

**Comment by isbm** [comment]
@cachedout @rallytime ping again. :wink:

File: salt/config/__init__.py (hunk 10 above)
Comment (cachedout) [reply salt/config/__init__.py]: I am thinking of this in the context of the `VALID_OPTS` dictionary only. I understand that in the _actual config_ that it is nested but it is not that way in `VALID_OPTS`.

Comment (isbm) [reply salt/config/__init__.py]: @cachedout OK, maybe I don't understand how to properly declare it then. As I understood, normally when it is defined:
```python
foo = (bar, baz)
```
So logically, `bar` and `baz` should be previously defined somewhere:
```python
bar = int
baz = boolean
```
Hence, valid construction is:
```yaml
foo = (123, True)
```

No?

Please just tell me then where should I place it the way that it doesn't get this ugly prefix, but stays just `port`? In this case seem `port` is _just_ generic port. And there is no `port` at the root level, so it supposed to be invalid, once defined, unless the whole config definition is just broken.

Comment (isbm) [reply salt/config/__init__.py]: @thatch45 could you please help here? Seem we stuck.

Comment (isbm) [reply salt/config/__init__.py]: @cachedout Ok, I looked at the `VALID_OPTS` dict once again. Does it mean that it doesn't even designed to represent a schema, but only a top-level? I thought that it kind of represents a structure of the config. But looking at this again, seems like I misunderstood. So you worry that:
```python
VALID_OPTS = {
    'port': int
}
```
...will bring up `port` at the root level of the config. Where I thought that if there will be _anywhere_ key e.g. `fridge` or `james_bond`, then the whole thing should at least `log.error('...` this that such key is invalid _anywhere_, no matter it is inside some structure or at the root level. And the same validation should also `log.error('...` if the key `port` is a string or float etc., regardless where it is placed.

Am I wrong?

Comment (isbm) [reply salt/config/__init__.py]: @cachedout that said, my reasoning was this:

- Define a key `port` in `VALID_OPTS` to make sure it is a) valid at all and b) it is `int` checked.
- Use it in my structure.

But if you afraid that somebody will start using `port` at the top level for something like `pancake_cooker_checker`, then yeah, this is wrong. But at this point I'd rather start thinking about the entire _schema_ of the Config, rather then just a one-level dictionary that defines what it defines: a valid options.

**Comment by isbm** [comment]
@cachedout @rallytime ping?

File: salt/config/__init__.py (hunk 10 above)
Comment (cachedout) [reply salt/config/__init__.py]: @isbm You are not wrong. The schema was not designed for this case and I agree with you that is a problem. You'll see other places where this has been worked around by prefixing the intended key with its parent and an underscore.

We should absolutely fix this in a separate PR.

If you'd like to go ahead with this as-is, I can live with it though.

**Review by cachedout — APPROVED** [review]

File: salt/config/__init__.py (hunk 10 above)
Comment (isbm) [reply salt/config/__init__.py]: Exactly. I would be happy doing this. For this matter I would say, there is another related problems: config is not really _validated_ before the Master/Minion starts. It only crashes on syntax error, but if somebody will add:

```yaml
foo: bar
```
This won't warning you something like "Option `foo` is unknown". But it gets even worse — nothing will happen if you do just this:
```yaml
ipv6: Take two
ret_port: "4506"
color: yes
```
At this point I see literally zero validation, and this should be really addressed properly. In a separate PR.

Comment (isbm) [reply salt/config/__init__.py]: @cachedout as of prefixes, they are ugly and messy. :wink: You don't say in English "pizza with pizzapepperoni" and "potato with potatopepperoni". You say "<object> with <another-object>". So why not just define "pepperoni" and serve with pizza or potato separately? Pizza can contain potato as well, BTW. This is the same reason why do we have 10 different boto_* modules and other module/function names that do almost the same thing, but called slightly differently. We had that discussion at SUSECon2017 about namespaces, remember? :wink:

**Comment by damon-atkins** [comment]
Does this need DNS Text field or something to have a finger print of the masters key, to prevent someone faking a master?

**Comment by isbm** [comment]
@damon-atkins [this explains](https://github.com/isbm/salt/blob/6c901aa902261553bf9d0056dc7e10a92ee13f8c/doc/topics/releases/oxygen.rst#limitations) your concerns.

**Comment by rallytime** [comment]
Go Go Jenkins!

**Comment by rallytime** [comment]
@isbm Can you rebase this one more time? There is a merge conflict. Then we should be able to get this in.

**Comment by rallytime** [comment]
Thanks @isbm!

**Comment by isbm** [comment]
Alt Text

**Pull request merged**